include/irr/asset/ICPUVirtualTexture.h

```diff
@@ -55,70 +55,109 @@ class ICPUVirtualTexture final : public IVirtualTexture<ICPUImageView>
+        copy.axisWraps[1] = _wrapv;
+        copy.axisWraps[2] = asset::ISampler::ETC_CLAMP_TO_EDGE;
+        copy.borderColor = asset::ISampler::ETBC_FLOAT_OPAQUE_BLACK;
+        copy.extent = params.extent;
+        copy.layerCount = 1u;
+        copy.inMipLevel = 0u;
+        copy.inOffset = {0u,0u,0u};
+        copy.inBaseLayer = 0u;
+        copy.outOffset = {0u,0u,0u};
+        copy.outBaseLayer = 0u;
+        copy.outMipLevel = 0u;
+        copy.paddedExtent = {paddedExtent,paddedExtent,1u};
+        copy.relativeOffset = {0u,0u,0u};
+        copy.inImage = _img;
+        copy.outImage = paddedImg.get();
+
+        asset::CPaddedCopyImageFilter::execute(&copy);
+
+        using mip_gen_filter_t = asset::CMipMapGenerationImageFilter<asset::CBoxImageFilterKernel,asset::CBoxImageFilterKernel>;
+        //generate all mip levels
+        {
+            mip_gen_filter_t::state_type genmips;
+            genmips.baseLayer = 0u;
+            genmips.layerCount = 1u;
+            genmips.startMipLevel = 1u;
+            genmips.endMipLevel = paddedImg->getCreationParameters().mipLevels;
+            genmips.inOutImage = paddedImg.get();
+            genmips.scratchMemoryByteSize = mip_gen_filter_t::getRequiredScratchByteSize(&genmips);
+            genmips.scratchMemory = reinterpret_cast<uint8_t*>(_IRR_ALIGNED_MALLOC(genmips.scratchMemoryByteSize,_IRR_SIMD_ALIGNMENT));
+            mip_gen_filter_t::execute(&genmips);
+            _IRR_ALIGNED_FREE(genmips.scratchMemory);
+        }
+
+        return std::make_pair(std::move(paddedImg), originalExtent);
+    }
+
     ICPUVirtualTexture(
         const base_t::IVTResidentStorage::SCreationParams* _residentStorageParams,
         uint32_t _residentStorageCount,
-        uint32_t _pgTabSzxy_log2 = 8u,
-        uint32_t _pgTabLayers = 256u,
         uint32_t _pgSzxy_log2 = 7u,
+        uint32_t _pgTabLayers = 32u,
         uint32_t _tilePadding = 9u,
         uint32_t _maxAllocatableTexSz_log2 = 14u
     ) : IVirtualTexture(
-        _pgTabSzxy_log2, _pgTabLayers, _pgSzxy_log2, _tilePadding
+        _maxAllocatableTexSz_log2-_pgSzxy_log2, _pgTabLayers, _pgSzxy_log2, _tilePadding
     ) {
-        m_pageTable = createPageTable(_pgTabSzxy_log2, _pgTabLayers, _pgSzxy_log2, _maxAllocatableTexSz_log2);
+        m_pageTable = createPageTable(m_pgtabSzxy_log2, _pgTabLayers, _pgSzxy_log2, _maxAllocatableTexSz_log2);
         initResidentStorage(_residentStorageParams, _residentStorageCount);
     }
 
-    STextureData pack(const ICPUImage* _img, const IImage::SSubresourceRange& _subres, ISampler::E_TEXTURE_CLAMP _wrapu, ISampler::E_TEXTURE_CLAMP _wrapv, ISampler::E_TEXTURE_BORDER_COLOR _borderColor) override
+    bool commit(const SMasterTextureData& _addr, const ICPUImage* _img, const VkExtent3D& _mip0extent, const IImage::SSubresourceRange& _subres, ISampler::E_TEXTURE_CLAMP _wrapu, ISampler::E_TEXTURE_CLAMP _wrapv, ISampler::E_TEXTURE_BORDER_COLOR _borderColor) override
```

Comment:
`const VkExtent3D& _mip0extent, const IImage::SSubresourceRange& _subres, ISampler::E_TEXTURE_CLAMP _wrapu, ISampler::E_TEXTURE_CLAMP _wrapv` are almost all in `SMasterTextureData` (except for baseLayer of `IImage::SSubresourceRange`)

Reply:
there's no base miplevel, base layer and miplevel count, so i'm leaving whole `_subres`